File: THCS/2022_2023/ThanhHoa_2223/Bai3/THUASONT.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

using namespace std;

void prime_factorization(long long N, vector<pair<long long, int>>& factors) {
    // Xử lý số 2
    int count = 0;
    while (N % 2 == 0) {
        count++;
        N /= 2;
    }
    if (count > 0) {
        factors.push_back({2, count});
    }

    // Xử lý các số lẻ từ 3 đến sqrt(N)
    for (long long i = 3; i <= sqrt(N); i += 2) {
        count = 0;
        while (N % i == 0) {
            count++;
            N /= i;
        }
        if (count > 0) {
            factors.push_back({i, count});
        }
    }

    // Nếu N là một số nguyên tố lớn hơn 2
    if (N > 2) {
        factors.push_back({N, 1});
    }
}

int main() {
    // Đọc dữ liệu từ file
    ifstream infile("THUASONT.INP");
    long long N;
    infile >> N;
    infile.close();

    // Lưu các thừa số nguyên tố và số lần xuất hiện của chúng
    vector<pair<long long, int>> factors;

    // Phân tích N thành các thừa số nguyên tố
    prime_factorization(N, factors);

    // Ghi kết quả vào file
    ofstream outfile("THUASONT.OUT");
    outfile << factors.size() << endl; // Số lượng thừa số nguyên tố
    for (const auto& factor : factors) {
        outfile << factor.first << " " << factor.second << endl; // p_i và a_i
    }
    outfile.close();

    return 0;
}

File: THCS/2022_2023/ThanhHoa_2223/Bai3/THUASONT.py
import math

def prime_factorization(N):
    factors = []
    # Xử lý số 2
    count = 0
    while N % 2 == 0:
        count += 1
        N //= 2
    if count > 0:
        factors.append((2, count))

    # Xử lý các số lẻ từ 3 đến sqrt(N)
    for i in range(3, int(math.sqrt(N)) + 1, 2):
        count = 0
        while N % i == 0:
            count += 1
            N //= i
        if count > 0:
            factors.append((i, count))

    # Nếu N là một số nguyên tố lớn hơn 2
    if N > 2:
        factors.append((N, 1))

    return factors

def main():
    # Đọc dữ liệu từ file
    with open("THUASONT.INP", "r") as infile:
        N = int(infile.readline().strip())

    # Phân tích N thành các thừa số nguyên tố
    factors = prime_factorization(N)

    # Ghi kết quả vào file
    with open("THUASONT.OUT", "w") as outfile:
        outfile.write(f"{len(factors)}\n")  # Số lượng thừa số nguyên tố
        for factor in factors:
            outfile.write(f"{factor[0]} {factor[1]}\n")  # p_i và a_i

if __name__ == "__main__":
    main()
